dosbox-staging/dosbox-staging
# Initial clipboard sharing support

# Description

Initial clipboard sharing support - see _Release notes_ for details. If secure mode is enabled, clipboard functions in _detached_ mode - appears to work, but there is no exchange with the host clipboard.


## Related issues

https://github.com/dosbox-staging/dosbox-staging/issues/1864


# Release notes

- Implemented command `CLIP.EXE` (originally available on the _Windows 98 Resource Kit_ CD, now part of modern _Microsoft Windows_ - [link](https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/clip)) to allow clipboard handling from within BAT scripts or command line.
- Pressing _Ctrl+V_ in DOSBox command line now pastes the first line of the clipboard's text content.

# Manual testing

- using your favorite (host side) ASCII editor, put something into system clipboard; from within DOSBox execute command `clip` or press Ctrl+V to test pasting
- execute command `dir | clip` or `clip < Z:\AUTOEXEC.BAT` within DOSBox - check that the result can be now pasted on the host side, or on the guest side (see previous point)

Don't be afraid to try national characters, or to use `keyb` command to switch code pages while the text is in the clipboard - example (RightAlt + letter produces a relevant Polish national character):

```BAT
keyb pl 668
echo ąęść | clip
clip
keyb pl 667
clip
keyb us 437
clip
keyb pl 667
clip
keyb pl 668
clip
```

# Checklist

I have:
